### ISSUE TITLE
Only one/few collators building blocks if parachain is underscheduled

### ISSUE THREAD

**Issue by alindima** [paritytech/polkadot-sdk#6667]
Assume using the slot-based collator and a parachain with slot duration of 2 seconds.
In order to fully take advantage of elastic scaling, the parachain should have 3 cores scheduled at all times.
But what happens if it doesn't, for some reason (by accident or just the parachain decides to scale down based on lower demand)?

In this case, only one collator will be authoring blocks, which is not good.
This is because:

Collator A builds the first block.
Collator B's slot kicks in, but collator A already built a block for this core. So go back to sleep.
Collator C's slot kicks in, but collator A already built a block for this core. So go back to sleep.
Collator A's slot kicks in, there's a new relay chain block and a new opportunity to build a new block.

In the end, only collator A will builds blocks.

This can be fixed by having some dynamic slot duration based on the amount of scheduled cores.

CC: @sandreim @skunert

**Comment by skunert** [comment]
This does not sound like something we need to solve. Your problem statement assumes that we only have 3 collators. As soon as the slot_duration / core_count / collator_count does not match up anymore, someone else will author (e.g. in your example, if we have collator E and G, someone else will author).

Same could happen if we have no elastic scaling but are only scheduled on a core every x blocks.

**Comment by alindima** [comment]
> This does not sound like something we need to solve. Your problem statement assumes that we only have 3 collators. As soon as the slot_duration / core_count / collator_count does not match up anymore, someone else will author (e.g. in your example, if we have collator D and E, someone else will author).

Indeed it depends on this correlation. But we don't have any restriction AFAIK on the exact number of collators a parachain should have.
Another example would be slot duration of 3 seconds, 1 scheduled core, 4 collators. Only A and B will author.

So basically it seems a problem if the number of expected cores is a divisor of the number of collators

> Same could happen if we have no elastic scaling but are only scheduled on a core every x blocks.

Agreed, but in my view this is another reason why this should be solved. Doesn't sounds too good if a core-sharing parachain that has one assignment every 5 blocks and 5 collators only uses one of the collators to author (which is a realistic scenario, right?). What if that one collator is down? Even if it's not, it will be reaping all of the block authoring rewards or could pick which transactions to include

**Comment by sandreim** [comment]
Yeah, low priority for now, but will become relevant once we can support multiple collators. 

As for the solution, I think some flavor of @eskimor's proposal in https://github.com/paritytech/polkadot-sdk/issues/4813 is a good fix for the problem. The gist of it relevant to this issue here is in the `Limitations/Consequences` section.

**Comment by eskimor** [comment]
Good catch and indeed one more reason to not decrease slot times too much, but instead let one collator build all the blocks for a given relay block.

**Comment by eskimor** [comment]
> What about the case with core-sharing? I think this needs a separate solution

Can be solved by #4813 too, you just need to replace `3` in rule 3 with something larger.
